src/tools/pbs_python.c

```diff
@@ -2484,6 +2484,7 @@ main(int argc, char *argv[], char *envp[])
 		} else {
 			rc = execve(python_path, argv, lenvp);
 		}
+		free(python_path);
```

Comment:
There are several ways you can return out of this function without freeing python_path

Reply:
I did not really get you. Since `pbs_asprintf()` won't free the variable `python_path`, I am making an explicit call here.

Reply:
@sujatapatnaik52 - @bhroam means to say that execve() returns only on failure. It DOES NOT return if it is successful.

Reply:
Thank you @prakashcv13. I got that. @bhroam @prakashcv13  I was trying a program to check the behaviour of execve. Below is the same:

```
[root@rhel8-test PBS-22435]# cat test_leak.c 
#include<stdio.h> 
#include<stdio.h> 

int main() 
{ 
    int i; 
    printf("I am test call for leak from the parent process"); 
    printf("\n");       
    return 0; 
} 
[root@rhel8-test PBS-22435]# gcc test_leak.c -o test_leak
[root@rhel8-test PBS-22435]# cat test.c 
#include <stdio.h>
#include <sys/types.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>

int main()
{
  char *path = "./test_leak";
  char *my_test = (char*)malloc(20 * sizeof(char));
  char *args[] = {NULL};
  char *env[] = {NULL};
  strcpy(my_test,path);
  int rc;
  rc = execve(my_test, args, env); 
  printf("I won't get printed... :-("); 
  return 0; 
} 

[root@rhel8-test PBS-22435]# gcc test.c -o test
[root@rhel8-test PBS-22435]# valgrind --trace-children=yes --leak-check=yes ./test
==42685== Memcheck, a memory error detector
==42685== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==42685== Using Valgrind-3.14.0 and LibVEX; rerun with -h for copyright info
==42685== Command: ./test
==42685== 
==42685== Memcheck, a memory error detector
==42685== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==42685== Using Valgrind-3.14.0 and LibVEX; rerun with -h for copyright info
==42685== Command: ./test_leak
==42685== 
I am test call for leak from the parent process
==42685== 
==42685== HEAP SUMMARY:
==42685==     in use at exit: 0 bytes in 0 blocks
==42685==   total heap usage: 1 allocs, 1 frees, 1,024 bytes allocated
==42685== 
==42685== All heap blocks were freed -- no leaks are possible
==42685== 
==42685== For counts of detected and suppressed errors, rerun with: -v
==42685== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)

### If I edit test.c to make the file name "test_leak" to just "test_le" i.e.
[root@rhel8-test PBS-22435]# cat test.c
#include <stdio.h>
#include <sys/types.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>

int main()
{
  char *path = "./test_le";
  char *my_test = (char*)malloc(20 * sizeof(char));
  char *args[] = {NULL};
  char *env[] = {NULL};
  strcpy(my_test,path);
  int rc;
  rc = execve(my_test, args, env); 
  printf("I won't get printed... :-("); 
  return 0; 
} 
### Valgrind output
[root@rhel8-test PBS-22435]# valgrind --trace-children=yes --leak-check=yes ./test
==42731== Memcheck, a memory error detector
==42731== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==42731== Using Valgrind-3.14.0 and LibVEX; rerun with -h for copyright info
==42731== Command: ./test
==42731== 
I won't get printed... :-(==42731== 
==42731== HEAP SUMMARY:
==42731==     in use at exit: 20 bytes in 1 blocks
==42731==   total heap usage: 2 allocs, 1 frees, 1,044 bytes allocated
==42731== 
==42731== 20 bytes in 1 blocks are definitely lost in loss record 1 of 1
==42731==    at 0x4C30E8B: malloc (vg_replace_malloc.c:309)
==42731==    by 0x40067F: main (in /home/sujata/PBS-22435/test)
==42731== 
==42731== LEAK SUMMARY:
==42731==    **definitely lost: 20 bytes in 1 blocks**
==42731==    indirectly lost: 0 bytes in 0 blocks
==42731==      possibly lost: 0 bytes in 0 blocks
==42731==    still reachable: 0 bytes in 0 blocks
==42731==         suppressed: 0 bytes in 0 blocks
==42731== 
==42731== For counts of detected and suppressed errors, rerun with: -v
==42731== ERROR SUMMARY: 1 errors from 1 contexts (suppressed: 0 from 0)

To the above program I added a free() to test.c and I see the following results:
[root@rhel8-test PBS-22435]# valgrind --trace-children=yes --leak-check=yes ./test
==42747== Memcheck, a memory error detector
==42747== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==42747== Using Valgrind-3.14.0 and LibVEX; rerun with -h for copyright info
==42747== Command: ./test
==42747== 
I won't get printed... :-(==42747== 
==42747== HEAP SUMMARY:
==42747==     in use at exit: 0 bytes in 0 blocks
==42747==   total heap usage: 2 allocs, 2 frees, 1,044 bytes allocated
==42747== 
==42747== All heap blocks were freed -- no leaks are possible
==42747== 
==42747== For counts of detected and suppressed errors, rerun with: -v
==42747== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```
Reference - https://stackoverflow.com/questions/3693335/what-happens-to-dynamic-allocated-memory-when-calling-execv

It appears to me that the dynamic memory is getting freed in the new process only if the execve call is successful and it does not get freed if it is not successful and maybe we have to explicitly free it. Let me know if my understanding is correct.

Reply:
The free after execve() seems right to me for handling failure case of execve().

From above discussion, If I understand correctly, the point is we would need to free the dynamically allocated memory referred by "python_path" pointer from several other places too. At these line numbers like 2370, 2375, 2383, 2404, 2437, 2477 I also see a return statement in this main function. It's better to add free(python_path); before those returns.

Reply:
Thank you @bremanandjk for confirming my understanding. Thank you for catching that. I will make the changes.

Reply:
yes, my assumption about @bhroam's comment was wrong - sorry for the confusion.

Reply:
Thanks @bremanandjk, that's exactly what I was trying to get at (and failing).